modules/gapi/include/opencv2/gapi/s11n.hpp

```diff
@@ -175,16 +189,130 @@ IIStream& operator>> (IIStream& is, std::vector<T> &ts) {
     return is;
 }
 
-namespace detail {
-    // Will be used along with default types if possible in specific cases (compile args, etc)
-    // Note: actual implementation is defined by user
-    template<typename T>
-    struct GAPI_EXPORTS S11N {
-        static void serialize(IOStream &, const T &) {}
-        static T deserialize(IIStream &) { T t; return t; }
-    };
+namespace detail
+{
+// Note: actual implementation is defined by user
+template<typename T>
+struct GAPI_EXPORTS S11N;
+
+
+template<typename T> struct wrap_serialize<T, cv::GCompileArg> {
+    static std::function<void(gapi::s11n::IOStream& os, const util::any& arg)> serialize;
+
+private:
+    template<typename> using sfinae_true = std::true_type;
+
+    template<typename Q = T>
+    static auto try_call_serialize(gapi::s11n::IOStream& os, const util::any& arg, int)
+        -> sfinae_true<decltype(S11N<Q>::serialize(os, util::any_cast<Q>(arg)), void())> {
+
+        S11N<Q>::serialize(os, util::any_cast<Q>(arg));
+        return sfinae_true<void>{};
+    }
+
+    // FIXME: Add compile-time check that this can't be called for custom types.
+    //        Such behaviour can alternatively be implemented in S11N<T> with std::enable_if
+    //        contraining T to be types defined by framework.
+    template<typename Q = T>
+    static auto try_call_serialize(gapi::s11n::IOStream& os, const util::any& arg, long)
+        -> sfinae_true<decltype(os << std::declval<const typename std::add_lvalue_reference<Q>::type>(), void())> {
+
+        os << util::any_cast<Q>(arg);
+        return sfinae_true<void>{};
+    }
+
+    template<typename Q = T>
+    static std::false_type try_call_serialize(gapi::s11n::IOStream &os, const util::any& arg, ...);
+
+    static void call_serialize(gapi::s11n::IOStream& os, const util::any& arg) {
+        try_call_serialize<T>(os, arg, 0);
+    }
+};
```

Comment:
No need in this whole SFINAE thing.

The only purpose of `wrap_serialize` is to produce an _untyped_ callback over a _typed_ function implemented by user. Take `meta_helper` or `call_helper` for example.

Not more than that, please.

Reply:
Reworked